File: Examples/Example-Java/app/src/main/java/com/zettle/payments/android/java_example/CardReaderActivity.java
package com.zettle.payments.android.java_example;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.SpannableStringBuilder;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.MutableLiveData;

import com.google.android.material.snackbar.Snackbar;
import com.zettle.sdk.feature.cardreader.payment.TippingStyle;
import com.zettle.sdk.feature.cardreader.payment.TransactionReference;
import com.zettle.sdk.feature.cardreader.ui.CardReaderAction;
import com.zettle.sdk.feature.cardreader.ui.payment.CardPaymentResult;
import com.zettle.sdk.feature.cardreader.ui.readers.CardReadersActivity;
import com.zettle.sdk.features.ActionUtils;
import com.zettle.sdk.features.Transaction;
import com.zettle.sdk.ui.ZettleResult;
import com.zettle.sdk.ui.ZettleResultKt;

import java.util.Objects;
import java.util.UUID;

public class CardReaderActivity extends AppCompatActivity {

    private Button chargeButton;
    private Button refundButton;
    private Button retrieveButton;
    private EditText refundAmountEditText;
    private Button settingsButton;
    private EditText amountEditText;
    private CheckBox installmentsCheckBox;
    private MutableLiveData<String> lastPaymentTraceId;
    private Button tippingStyleButton;
    private TippingStyle tippingStyle = TippingStyle.None;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_card_reader);
        chargeButton = findViewById(R.id.charge_btn);
        refundButton = findViewById(R.id.refund_btn);
        retrieveButton = findViewById(R.id.retrieve_btn);
        settingsButton = findViewById(R.id.settings_btn);
        amountEditText = findViewById(R.id.amount_input);
        refundAmountEditText = findViewById(R.id.refund_amount_input);
        tippingStyleButton = findViewById(R.id.tipping_style_btn);
        installmentsCheckBox = findViewById(R.id.installments_check_box);
        lastPaymentTraceId = new MutableLiveData<>();

        lastPaymentTraceId.observe(this, value -> {
            refundButton.setEnabled(value != null);
            retrieveButton.setEnabled(value != null);
        });

        chargeButton.setOnClickListener(v -> onChargeClicked());
        refundButton.setOnClickListener(v -> onRefundClicked());
        retrieveButton.setOnClickListener(v -> onRetrieveClicked());
        settingsButton.setOnClickListener(v -> onSettingsClicked());
        tippingStyleButton.setOnClickListener(v -> onTippingStyleClicked());

        setTippingStyleTitle();
        getSupportFragmentManager().setFragmentResultListener(TippingStyleBottomSheet.REQUEST_KEY, this, (requestKey, result) -> {
            TippingStyle newTippingStyle = (TippingStyle) result.getSerializable(TippingStyleBottomSheet.TIPPING_STYLE_KEY);
            tippingStyle = newTippingStyle != null ? newTippingStyle : TippingStyle.None;
            setTippingStyleTitle();
        });
    }

    private final ActivityResultLauncher<Intent> paymentLauncher = registerForActivityResult(new StartActivityForResult(), result -> {
        if (result.getResultCode() == Activity.RESULT_OK && result.getData() != null) {
            ZettleResult parsed = ZettleResultKt.zettleResult(result.getData());
            if (parsed instanceof ZettleResult.Completed) {
                showSnackBar("Payment completed");
                CardPaymentResult.Completed casted = (CardPaymentResult.Completed) parsed;
                lastPaymentTraceId.setValue(Objects.requireNonNull(casted.getPayload().getReference()).getId());
                refundAmountEditText.setText(new SpannableStringBuilder()
                        .append(String.valueOf(casted.getPayload().getAmount())));
            } else if (parsed instanceof ZettleResult.Failed) {
                showSnackBar("Payment failed " + ((ZettleResult.Failed) parsed).getReason());
            } else if (parsed instanceof ZettleResult.Cancelled) {
                showSnackBar("Payment canceled");
            }
        }
    });

    private final ActivityResultLauncher<Intent> refundLauncher = registerForActivityResult(new StartActivityForResult(), result -> {
        if (result.getResultCode() == Activity.RESULT_OK && result.getData() != null) {
            ZettleResult parsed = ZettleResultKt.zettleResult(result.getData());
            if (parsed instanceof ZettleResult.Completed) {
                showSnackBar("Refund completed");
            } else if (parsed instanceof ZettleResult.Failed) {
                showSnackBar("Refund failed " + ((ZettleResult.Failed) parsed).getReason());
            } else if (parsed instanceof ZettleResult.Cancelled) {
                showSnackBar("Refund canceled");
            }
        }
    });

    private void showSnackBar(String text) {
        ViewGroup viewGroup = findViewById(android.R.id.content);
        Snackbar.make(viewGroup.getChildAt(0), text, Snackbar.LENGTH_LONG).show();
    }

    private void onChargeClicked() {
        String amountEditTextContent = amountEditText.getText().toString();
        if (amountEditTextContent.isEmpty()) {
            return;
        }
        String internalTraceId = UUID.randomUUID().toString();
        long amount = Utils.parseLong(amountEditText.getText());
        boolean enableInstallments = installmentsCheckBox.isChecked();

        TransactionReference reference = new TransactionReference.Builder(internalTraceId)
                .put("PAYMENT_EXTRA_INFO", "Started from home screen")
                .build();

        CardReaderAction.Payment payment = new CardReaderAction.Payment(
                reference, amount, tippingStyle, enableInstallments
        );
        Intent intent = ActionUtils.charge(payment, this);
        paymentLauncher.launch(intent);
    }

    private void onSettingsClicked() {
        startActivity(CardReadersActivity.newIntent(this));
    }

    private void onTippingStyleClicked() {
        TippingStyleBottomSheet.newInstance().show(getSupportFragmentManager(), TippingStyleBottomSheet.class.getSimpleName());
    }

    private void setTippingStyleTitle() {
        String tippingStyleTitle = "Tipping Style - " + tippingStyle.name();
        tippingStyleButton.setText(tippingStyleTitle);
    }

    private void onRefundClicked() {
        String internalTraceId = lastPaymentTraceId.getValue();
        boolean isDevMode = ((MainApplication) getApplication()).isDevMode();

        if (internalTraceId == null && !isDevMode) {
            showSnackBar("No payment taken");
            return;
        }

        Long amount = Utils.parseLong(refundAmountEditText.getText());

        TransactionReference reference = new TransactionReference.Builder(UUID.randomUUID().toString())
                .put("REFUND_EXTRA_INFO", "Started from home screen")
                .build();

        CardReaderAction.Refund refund = new CardReaderAction.Refund(
                reference, amount, internalTraceId, null, null
        );
        refundLauncher.launch(ActionUtils.refund(refund,this));
    }

    private void onRetrieveClicked() {
        String internalTraceId = lastPaymentTraceId.getValue();
        boolean isDevMode = ((MainApplication) getApplication()).isDevMode();

        if (internalTraceId == null && !isDevMode) {
            showSnackBar("No payment taken");
            return;
        }

        Transaction transaction = new CardReaderAction.Transaction(internalTraceId);
        ActionUtils.retrieve(transaction, result -> {
            if (result instanceof ZettleResult.Completed) {
                showSnackBar("Retrieve payment completed");
            } else if (result instanceof ZettleResult.Failed) {
                showSnackBar("Retrieve failed " + ((ZettleResult.Failed) result).getReason());
            } else if (result instanceof ZettleResult.Cancelled) {
                showSnackBar("Retrieve canceled");
            }
            return null;
        });
    }

}

File: Examples/Example-Kotlin/app/src/main/java/com/zettle/payments/android/kotlin_example/CardReaderActivity.kt
package com.zettle.payments.android.kotlin_example

import android.app.Activity
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Parcelable
import android.text.SpannableStringBuilder
import android.view.ViewGroup
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.TextView
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.MutableLiveData
import com.google.android.material.snackbar.Snackbar
import com.zettle.sdk.feature.cardreader.payment.PayPalReaderTippingStyle
import com.zettle.sdk.feature.cardreader.payment.TippingConfiguration
import com.zettle.sdk.feature.cardreader.payment.TippingPercentageOptions
import com.zettle.sdk.feature.cardreader.payment.Transaction
import com.zettle.sdk.feature.cardreader.payment.TransactionReference
import com.zettle.sdk.feature.cardreader.payment.ZettleReaderTippingStyle
import com.zettle.sdk.feature.cardreader.payment.refunds.CardPaymentPayload
import com.zettle.sdk.feature.cardreader.payment.refunds.RefundPayload
import com.zettle.sdk.feature.cardreader.ui.CardReaderAction
import com.zettle.sdk.feature.cardreader.ui.RetrieveResult
import com.zettle.sdk.feature.cardreader.ui.payment.CardPaymentResult
import com.zettle.sdk.feature.cardreader.ui.refunds.RefundResult
import com.zettle.sdk.features.charge
import com.zettle.sdk.features.refund
import com.zettle.sdk.features.retrieve
import com.zettle.sdk.features.show
import com.zettle.sdk.ui.ZettleResult
import com.zettle.sdk.ui.zettleResult
import java.util.*
import kotlin.math.abs

class CardReaderActivity : AppCompatActivity() {

    private lateinit var chargeButton: Button
    private lateinit var refundButton: Button
    private lateinit var retrieveButton: Button
    private lateinit var refundAmountEditText: EditText
    private lateinit var settingsButton: Button
    private lateinit var amountEditText: EditText
    private lateinit var installmentsCheckBox: CheckBox
    private lateinit var lastPaymentTraceId: MutableLiveData<String?>

    private lateinit var selectTippingStyleButton: Button
    private lateinit var ztrTippingStyleLabel: TextView
    private lateinit var pprTippingStyleLabel: TextView
    private lateinit var openPprTippingSettingsButton: Button

    private var tippingConfiguration: TippingConfiguration? = TippingConfiguration(
        ZettleReaderTippingStyle.None,
        PayPalReaderTippingStyle.None
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_card_reader)
        chargeButton = findViewById(R.id.charge_btn)
        refundButton = findViewById(R.id.refund_btn)
        retrieveButton = findViewById(R.id.retrieve_btn)
        settingsButton = findViewById(R.id.settings_btn)
        amountEditText = findViewById(R.id.amount_input)
        refundAmountEditText = findViewById(R.id.refund_amount_input)
        installmentsCheckBox = findViewById(R.id.installments_check_box)

        selectTippingStyleButton = findViewById(R.id.select_tipping_style_btn)
        selectTippingStyleButton.setOnClickListener { selectTippingStyle() }
        ztrTippingStyleLabel = findViewById(R.id.ztr_tipping_style_label)
        pprTippingStyleLabel = findViewById(R.id.ppr_tipping_style_label)
        openPprTippingSettingsButton = findViewById(R.id.open_ppr_tipping_settings_btn)
        openPprTippingSettingsButton.setOnClickListener { onTippingSettingsClicked() }

        lastPaymentTraceId = MutableLiveData()
        lastPaymentTraceId.observe(this) { value: String? ->
            refundButton.isEnabled = value != null
            retrieveButton.isEnabled = value != null
        }

        chargeButton.setOnClickListener { onChargeClicked() }
        refundButton.setOnClickListener { onRefundClicked() }
        retrieveButton.setOnClickListener { onRetrieveClicked() }
        settingsButton.setOnClickListener { onSettingsClicked() }

        setTippingStyleTitle()
    }

    private val paymentLauncher =
        registerForActivityResult(StartActivityForResult()) { activityResult ->
            if (activityResult.resultCode != Activity.RESULT_OK) {
                return@registerForActivityResult
            }

            when(val result = activityResult.data?.zettleResult()) {
                is ZettleResult.Completed<*> -> {
                    val payment: CardPaymentResult.Completed = CardReaderAction.fromPaymentResult(result)
                    showResultSheet(payment.payload.toPaymentResultData())
                    lastPaymentTraceId.value = payment.payload.reference?.id
                    refundAmountEditText.text = SpannableStringBuilder().append(payment.payload.amount.toString())
                }
                is ZettleResult.Cancelled -> showSnackBar("Payment canceled")
                is ZettleResult.Failed -> showSnackBar("Payment failed Reason#${result.reason.javaClass.simpleName}")
                null -> showSnackBar("Problem... null")
            }
        }

    private val refundLauncher =
        registerForActivityResult(StartActivityForResult()) { activityResult ->
            if (activityResult.resultCode != Activity.RESULT_OK) {
                return@registerForActivityResult
            }

            when(val result = activityResult.data?.zettleResult()) {
                is ZettleResult.Completed<*> -> {
                    val refund : RefundResult.Completed = CardReaderAction.fromRefundResult(result)
                    showResultSheet(refund.payload.toPaymentResultData())
                }
                is ZettleResult.Cancelled -> showSnackBar("Refund canceled")
                is ZettleResult.Failed -> showSnackBar("Refund failed Reason#${result.reason}")
                null -> showSnackBar("Problem... null")
            }
        }

    private fun Transaction.ResultPayload.toPaymentResultData() = PaymentResultData(
        title = "Card Payment",
        amount = formatPaymentAmount(amount),
        reference = reference?.id,
        resultList = toResultListItems()
    )

    private fun RefundPayload.toPaymentResultData() = PaymentResultData(
        title = "Card Refund",
        amount = formatPaymentAmount(-1 * abs(refundedAmount)),
        reference = null,
        resultList = toResultListItems(),
        resultType = ResultType.REFUND
    )

    private fun CardPaymentPayload.toPaymentResultData() = PaymentResultData(
        title = "Card Payment",
        amount = formatPaymentAmount(amount),
        reference = referenceId,
        resultList = toResultListItems()
    )

    private fun showResultSheet(resultData: PaymentResultData) {
        PaymentResultBottomSheet.newInstance(resultData)
            .show(supportFragmentManager, PaymentResultBottomSheet.TAG)
    }

    private fun showSnackBar(text: String) {
        findViewById<ViewGroup>(android.R.id.content).getChildAt(0).run {
            Snackbar.make(this, text, Snackbar.LENGTH_LONG).show()
        }
    }

    private fun onChargeClicked() {
        val amountEditTextContent = amountEditText.text.toString()
        if (amountEditTextContent == "") {
            return
        }
        val internalTraceId = UUID.randomUUID().toString()
        val amount = amountEditTextContent.toLong()
        val enableInstallments = installmentsCheckBox.isChecked
        val reference = TransactionReference.Builder(internalTraceId)
            .put("PAYMENT_EXTRA_INFO", "Started from home screen")
            .build()

        val intent: Intent = CardReaderAction.Payment(
            reference = reference,
            amount = amount,
            tippingConfiguration = tippingConfiguration,
            enableInstallments = enableInstallments
        ).charge(this)

        paymentLauncher.launch(intent)
    }

    private fun onRefundClicked() {
        val amount = refundAmountEditText.text.toLong() ?: 0L
        val isDevMode = (application as MainApplication).isDevMode

        if (lastPaymentTraceId.value == null && !isDevMode) {
            showSnackBar("No payment taken")
            return
        }

        val internalTraceId = lastPaymentTraceId.value ?: ""
        val reference = TransactionReference.Builder(UUID.randomUUID().toString())
            .put("REFUND_EXTRA_INFO", "Started from home screen")
            .build()

        val intent = CardReaderAction.Refund(
            amount = amount,
            paymentReferenceId = internalTraceId,
            refundReference = reference
        ).refund(this)
        refundLauncher.launch(intent)
    }

    private fun onRetrieveClicked() {
        val isDevMode = (application as MainApplication).isDevMode
        if (lastPaymentTraceId.value == null && !isDevMode) {
            showSnackBar("No payment taken")
            return
        }
        val internalTraceId = lastPaymentTraceId.value ?: ""
        CardReaderAction.Transaction(internalTraceId).retrieve {
            when (it) {
                is ZettleResult.Completed<*> -> {
                    val transaction: RetrieveResult.Completed = CardReaderAction.fromRetrieveTransactionResult(it)
                    showResultSheet(transaction.payload.toPaymentResultData())
                }
                is ZettleResult.Cancelled -> showSnackBar("Retrieve canceled")
                is ZettleResult.Failed -> showSnackBar("Retrieve failed Reason#${it.reason}")
            }
        }
    }

    private fun onSettingsClicked() {
        startActivity(CardReaderAction.Settings.show(this))
    }

    private val selectTippingStyleLauncher =
        registerForActivityResult(StartActivityForResult()) { result ->
            if (result.resultCode != Activity.RESULT_OK) {
                return@registerForActivityResult
            }

            val data = result.data ?: return@registerForActivityResult
            data.parcelableExtra<TippingConfiguration>(
                SelectTippingStyleActivity.TIPPING_CONFIGS_KEY
            )?.let {
                tippingConfiguration = it
                setTippingStyleTitle()
            }
        }

    private fun selectTippingStyle() {
        val intent = Intent(this, SelectTippingStyleActivity::class.java)
        intent.putExtra(SelectTippingStyleActivity.TIPPING_CONFIGS_KEY, tippingConfiguration)
        selectTippingStyleLauncher.launch(intent)
    }

    private fun setTippingStyleTitle() {
        tippingConfiguration?.let {
            ztrTippingStyleLabel.text = "Zettle Reader: ${it.zettleReaderTippingStyle.name}"
            val pprStyle = when (val style = it.payPalReaderTippingStyle) {
                is PayPalReaderTippingStyle.PredefinedPercentage -> style.javaClass.simpleName + style.options.displayText()
                else -> style.javaClass.simpleName
            }
            pprTippingStyleLabel.text = "PayPal Reader: $pprStyle"
        }
    }

    private fun onTippingSettingsClicked() {
        val intent = CardReaderAction.TippingSettings().show(this)
        startActivity(intent)
    }

    private fun TippingPercentageOptions?.displayText() = when (this) {
        null -> "(Options = Null)"
        else -> "($option1, $option2, $option3)"
    }
}

inline fun <reified T : Parcelable> Intent.parcelableExtra(key: String): T? = when {
    Build.VERSION.SDK_INT >= 33 -> getParcelableExtra(key, T::class.java)
    else -> @Suppress("DEPRECATION") getParcelableExtra(key) as? T
}